### PR TITLE
[SPARK-36266][SHUFFLE] Rename classes in shuffle RPC used for block push operations

### PR DESCRIPTION
**Pull request by Victsm**
### What changes were proposed in this pull request?
This is a follow-up to #29855 according to the [comments](https://github.com/apache/spark/pull/29855/files#r505536514)
In this PR, the following changes are made:

1. A new `BlockPushingListener` interface is created specifically for block push. The existing `BlockFetchingListener` interface is left as is, since it might be used by external shuffle solutions. These 2 interfaces are unified under `BlockTransferListener` to enable code reuse.
2. `RetryingBlockFetcher`, `BlockFetchStarter`, and `RetryingBlockFetchListener` are renamed to `RetryingBlockTransferor`, `BlockTransferStarter`, and `RetryingBlockTransferListener` respectively. This makes their names more generic to be reused across both block fetch and push.
3. Comments in `OneForOneBlockPusher` are further clarified to better explain how we handle retries for block push.

### Why are the changes needed?
To make code cleaner without sacrificing backward compatibility.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests.

### CODE REVIEW

**Comment by Victsm** [comment]
This PR picks up an earlier stale version of the same change as a follow up to the original push protocol RPC patch.
cc @mridulm @otterc @Ngone51

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala:116]: Add a precondition to make it explicit that we expect `BlockFetchingListener` ?

For all cases of casts we are making explicitly, can we add preconditions to ensure we do the type check upfront.

**Review by mridulm — COMMENTED** [review]
Took a quick pass.
Will wait for @Ngone51, @attilapiros, @dongjoon-hyun, @tgravescs, @jiangxb1987 and @otterc to also take a pass since they all reviewed the original PR.

**Comment by Victsm** [comment]
Test failure is unrelated to this change.

**Comment by Ngone51** [comment]
I haven't taken a deep look into code but just wondering according to the PR description. I know some code was added in 3.1 but push-based wasn't usable at that time. So, do we have to handle the compatibility issue? 

I also noticed a similar issue in another PR and @mridulm told me these kinds of cases don't have to be a compatibility issue.

**Comment by attilapiros** [comment]
I see why the adapter pattern is used here over simple renames. You also mentioned the reason: "without sacrificing backward compatibility" but could you explain the reason a little bit more in the description to have this information for the future reference.

**Comment by attilapiros** [comment]
@Ngone51 on compatibility issue it is meant keeping `BlockFetchingListener` as it is. As the `BlockFetchingListener` can be used by external shuffle solutions. @Victsm correct me if I am wrong.
By the way this is what I am missing from the description.

**Comment by Victsm** [comment]
@attilapiros you are right. The backward compatibility is meant to keep the existing `BlockFetchingListener` as is. Will clarify this better in the description.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:27]: Nit: Would prefer `BlockPushListener`. I think right now it is called `BlockPushing...` because the other one is called `BlockFetching...`? 
However, `BlockPushListener` sounds better and consistent with `BlockTransferListener`.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java:103]: nit:  `... but got a different listener`. If in future another type of listener is added then we will not need to change this error message here.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:122]: Right now this could be either be "fetch" or "push". 
I think we have introduced the `BlockTransferListener` which could now also be extended for some future use-case. 
So I am wondering whether this `transferTerm` should be passed in the constructor instead of assuming that it could be just push or fetch.
Another option would be to add a method to `BlockTransferListener`, let's say, `getLogPrefix` which by default can return `transfer`, but `BlockFetchingListener` can override it to `fetch` and `BlockPushListener` can override it to `push`.

Also, at other places in the comments, we are just assuming push or fetch. Maybe we just say "transfer" in the comments.

**Review by otterc — COMMENTED** [review]
Just have few comments.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java
Comment (Ngone51) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java:103]: maybe: `... but got listener inputListener`

(it's always good to show the unexpected value)

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java:147]: ditto

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:66]: nit: prefix with the `transferTerm`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:46]: Since this class has been renamed with the general suffix - `Transferor`, I think we can just say `... fist transfer for all blocks ...` here but hide the specific transfer type.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:112]: How about `transferType`?

**Comment by Ngone51** [comment]
BTW, I think it's better to use a separate JIRA id for this PR instead of a follow-up (which usually has less change).

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (attilapiros) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:43]: ..but transfer itself means a different thing.

I searched for synonyms and found this:

```
Synonyms: transferor

Definition: (law) someone who conveys a title or property to another

Similar words: transferer, transferrer

Definition: someone who transfers something
```
 
 Moreover this:

```
A person or thing that carries, holds, or conveys something

carrier bearer transporter conveyor courier haulier porter runner carman carter exporter hauler importer messenger mover shipper delivery man delivery woman delivery person delivery service letter carrier mail carrier shipping agent assistant sender merchant consigner caddie
```

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Victsm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:66]: Since this is a static final variable, would have to keep its naming generic.

**Comment by Victsm** [comment]
Created a separate ticket for this change under SPARK-36266 and addressed the additional comments.
@Ngone51 @attilapiros @otterc , please take another look.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (otterc) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:27]: Nit: Are we keeping `BlockPushingListener`? My suggestion was to call it `BlockPushListener`.

**Review by otterc — COMMENTED** [review]
Looks good to me. Just a nit. Thanks for working on this @Victsm

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (Ngone51) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:27]: I think `BlockPushingListener` is more consistent with `BlockFetchingListener.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:66]: ok

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:27]: Yeah, but `BlockPushListener` sounds better and is consistent with `BlockTransferListener`. Not a big deal though.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:43]: This's also public, although I have no idea how end-users could use it directly.
Shall we mark it as private from now? cc @mridulm

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java
Comment (Ngone51) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java:225]: `retry the transfer`?

**Comment by Ngone51** [comment]
LGTM if tests pass.

**Comment by Victsm** [comment]
The tests seem to have been abnormally disrupted based on what I can see from the log.
Not sure what's causing this.
cc @mridulm

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (Victsm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:27]: `BlockPushingListener` is supposed to be a peer to `BlockFetchingListener`. Their names being consistent feels more appropriate.

**Comment by Ngone51** [comment]
ok to test

**Review by attilapiros — COMMENTED** [review]
@Victsm I suspect the merged code which is used for running the unittests against introduces a build error:
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/core/src/test/scala/org/apache/spark/shuffle/ShuffleBlockPusherSuite.scala:335:76: not found: type BlockFetchingListener
[error]         val blockFetchListener = invocation.getArguments()(4).asInstanceOf[BlockFetchingListener]
[error]                                                                            ^
[error] one error found
```

Could you please rebase/merge this on/with top of the master and check the `ShuffleBlockPusherSuite.scala` after the rebase/merge?

**Comment by attilapiros** [comment]
Now I am sure. Line 335 in master:
https://github.com/apache/spark/blob/70a15868fc97e2b86c5ecc7bcf812bfdb05d98ea/core/src/test/scala/org/apache/spark/shuffle/ShuffleBlockPusherSuite.scala#L335

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockFetchingListener.java
Comment (attilapiros) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockFetchingListener.java:23]: I do not think the `transferType` (public static) members which only used  in `getTransferType()` methods are needed at all.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java
Comment (attilapiros) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockPushingListener.java:28]: https://github.com/apache/spark/pull/33340/files#r676015065

**Review by attilapiros — COMMENTED** [review]
Conditional LGTM: some nits (not blocking) but it must be rebased/merged onto/with master and the test should pass.

**Comment by mridulm** [comment]
Thanks for the reviews @Ngone51, @attilapiros, @otterc !
@Victsm Can you address @attilapiros's comments please ?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockFetchingListener.java
Comment (Victsm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockFetchingListener.java:23]: Intention was to reuse the same string in a straightforward way when `RetryingBlockTransferor` repeatedly invokes `getTransferType()` on multiple push/fetch listeners during block transfer retry.

Comment (attilapiros) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockFetchingListener.java:23]: The `transferType` now member val referencing the string literal "fetch" and when you use a string literal directly (i.e. "fetch") in the method (as that is the only place where it used) there should be no difference at all in performance. 

Like this

```
 @Override
  default String getTransferType() {
    return "fetch";
  }
```

It just works as fine but without an unnecessary member.

**Review by mridulm — APPROVED** [review]
LGTM, pending tests

**Comment by mridulm** [comment]
Merged to master and branch-3.2, +CC @gengliangwang
Thanks for working on this @Victsm.
Thanks for the reviews @Ngone51, @attilapiros, @otterc !

**Pull request closed without merging**
